Feature: Clevo/Tuxedo me unlock

SergiiDmytruk opened the pull request:
This is follow up on [Enabling support for ME unlock](https://groups.google.com/g/fwupd/c/e6yE6xy5Hjg) discussion on ML.

What it's needed in the end:
 1. Reading lock status from HSFS and sharing it for two use cases:
    1. State verification before unlocking.
    2. State verification before flashing ME firmware.
 2. Ability to unlock ME.
 3. Ability to flash ME firmware.

How it's done in this PR in its WIP state:
 1. Sharing ME lock status via `FuContext`.  Value is set by `intel-spi` plugin.  Consumed in `superio` and `flashrom`.
 2. Via `unlock` command for IT55 `superio` device, because unlocking is done through EC.
 3. Happens in `flashrom` plugin.  Device for ME firmware is added if ME region is unlocked.

Things to consider:
 1. What's the best way of presenting ME flashing to the user?  I guess ideally unlocking and flashing should be done for the same device. I saw that plugins can monitor created devices, but I think they can't invoke code from one another, so not sure how to implement all functionality in one place.
 2. `intel-spi` is re-used to not duplicate neither code nor quirks that are there and also because some kind of sharing would be necessary anyway.

ME device is incomplete at the moment, despite that the changes should capture the end goal relatively well. Please take a look to see if such changes are sane and if there are better ways of going about them.

P.S. First two commits are trivial changes for things I've noticed when going through the code and can be cherry-picked right away.

hughsie commented on plugins/flashrom/fu-flashrom-device.c:
```diff
@@ -309,6 +312,38 @@ fu_flashrom_device_constructed(GObject *obj)
+}
+
 static void
 fu_flashrom_device_finalize(GObject *object)
```
don't you need a `g_free(self->region);` here too?

SergiiDmytruk replied:
Yes, thanks. And we need `g_free()` in  `fu_superio_device_finalize()` too. The lack of `g_free()` there confused me.

hughsie commented on plugins/intel-spi/fu-intel-spi-device.c:
```diff
@@ -358,6 +362,11 @@ fu_intel_spi_device_setup(FuDevice *device, GError **error)
+	/* publish HSFS value for anyone else to use */
+	me_is_locked = !(self->hsfs & (1 << HSFS_FDV_OFF))     /* assume locked if not valid */
+	             || (self->hsfs & (1 << HSFS_FDOPSS_OFF)); /* use status bit if valid */
+	fu_context_set_me_lock_status(fu_device_get_context(device), me_is_locked);
```
rather than using the context, could we use something like https://github.com/fwupd/fwupd/blob/main/libfwupdplugin/fu-device-metadata.h#L10 and then get access to the device metadata using `vfuncs->device_registered` in the other plugins? It seems to me this is more of the state of a specific device rather than a whole system attribute, although I can certainty argue it the other way around too.

superm1 replied:
You can look at what plugin the device comes from.  There should only be one device from the intel-spi plugin, right?

SergiiDmytruk replied:
> There should only be one device from the intel-spi plugin, right?

I think it's one device but with one or more children, one of which actually corresponds to ME entry in IFD if it's present.

superm1 replied:
Then my proposal was that you either add that attribute to the top level device and just look for the top level device from the plugin or you add it to a child and check for a device that has the metadata set.
If you end up needing a reference again in the plugin, you can always cache it locally in the plugin when you find it.

SergiiDmytruk replied:
Resolved last hardware issue and can now test it properly. The child device even shows up in `get-devices --show-all-devices` as "Intel Management Engine", so I'll mark it with metadata. Thanks.

superm1 replied:
Another thing you can do is just use the locked device flag. Then no need to define a new metadata string.

superm1 commented on libfwupdplugin/fu-context.c:
```diff
@@ -32,6 +32,7 @@ typedef struct {
 	FuLidState lid_state;
 	guint battery_level;
 	guint battery_threshold;
+	gboolean me_locked;
```
ME is only specific to Intel - this shouldn't be in the context for all of FuContextPrivate.

superm1 commented on libfwupdplugin/fu-context.c:
```diff
@@ -888,6 +889,20 @@ fu_context_class_init(FuContextClass *klass)
+				     NULL,
+				     FALSE,
+				     G_PARAM_READWRITE | G_PARAM_STATIC_NAME);
+	g_object_class_install_property(object_class, PROP_BATTERY_STATE, pspec);
```
OK, I think review is trending it shouldn't be a context property anyway but if it does end up to stay in the context drop it here unless we think a client will need it.

superm1 commented on plugins/flashrom/fu-plugin-flashrom.c:
```diff
@@ -143,32 +145,65 @@ fu_plugin_flashrom_device_set_hwids(FuPlugin *plugin, FuDevice *device)
+			return FALSE;
+	} else if (g_strcmp0(region, "me") == 0) {
+		// XXX: should there be a hierarchy of devices?
+		// XXX: always create device but mark as FWUPD_DEVICE_FLAG_LOCKED if ME is locked?
```
I think you need to use `FWUPD_DEVICE_FLAG_LOCKED` somwehere in this PR.  Probably on whatever device is to be locked/unlocked.

hughsie commented on plugins/superio/fu-superio-it55-device.c:
```diff
@@ -47,6 +48,7 @@ struct _FuEcIt55Device {
 	FuSuperioDevice parent_instance;
 	gchar *prj_name;
 	AutoloadAction autoload_action;
+	gboolean me_locked;
```
I think you need to print this in the to_string() vfunc -- e.g. https://github.com/fwupd/fwupd/blob/main/plugins/cfu/fu-cfu-device.c#L27

hughsie commented on plugins/flashrom/fu-plugin-flashrom.c:
```diff
@@ -143,15 +151,21 @@ fu_plugin_flashrom_device_set_hwids(FuPlugin *plugin, FuDevice *device)
 	FuContext *ctx = fu_plugin_get_context(plugin);
 	const gchar *dmi_vendor;
-	g_autoptr(FuDevice) device = fu_flashrom_device_new(ctx);
+
```
no blank line here please

hughsie commented on plugins/flashrom/fu-plugin-flashrom.c:
```diff
@@ -143,15 +151,21 @@ fu_plugin_flashrom_device_set_hwids(FuPlugin *plugin, FuDevice *device)
 	fu_device_set_name(device, fu_context_get_hwid_value(ctx, FU_HWIDS_KEY_PRODUCT_NAME));
 	fu_device_set_vendor(device, fu_context_get_hwid_value(ctx, FU_HWIDS_KEY_MANUFACTURER));
 
+	instance_id = g_strdup_printf("IFD:%s", fu_ifd_region_to_string(region));
+	fu_device_add_instance_id(device, instance_id);
```
This isn't really an instance ID. What's here's already (i.e. what's the output of `sudo fwupdtool get-devices --plugins flashrom`)? Could you do something like `fu_device_add_instance_str(device, "REGION", fu_ifd_region_to_string(region));` and then `fu_device_build_instance_id(device, NULL, "FLASHROM", "VENDOR", "PRODUCT", "REGION", NULL);` -- We might also have to define the VENDOR and PRODUCT with `fu_device_add_instance_str` the same way -- and then we need it defined in the plugins/flashrom/README.md file....

SergiiDmytruk replied:
> This isn't really an instance ID.

Right, sorry, I'm confused about numerous IDs. Read relevant section of the tutorial and sources to improve my understanding.

> This isn't really an instance ID. What's here's already (i.e. what's the output of `sudo fwupdtool get-devices --plugins flashrom`)?

If I update `flashrom.quirk`, I get:
```
$ sudo fwupdtool get-devices --plugins flashrom
NS50MU
│
└─NS50MU:
      Device ID:          84bf22110f952e8493cc6ed99039b756d4129b5c
      Current version:    1.0
      Vendor:             Notebook (DMI:3mdeb Embedded Systems Consulting)
      GUIDs:              230c8b18-8d9b-53ec-838b-6cfc0383493a _ main-system-firmware
                          dc0de287-201b-5240-b8d2-0e5cead707ea _ Notebook&Not Applicable&NS50MU&Not Applicable&Notebook&NS50MU
                          8da9fe2b-90f4-5028-9bd9-c692b107d4ea _ Notebook&Not Applicable&NS50MU&Not Applicable
                          0039fec6-9ecf-5921-9961-905e148e052a _ Notebook&Not Applicable&NS50MU
                          6b26127f-384f-5ecc-9e94-02359da05f54 _ Notebook&Not Applicable&Notebook&NS50MU
                          58b02b86-65a5-56b6-bf24-fd9767043c58 _ Notebook&Not Applicable&NS50MU&Not Applicable&3mdeb Embedded Systems Consulting
      Device Flags:       _ Internal device
                          _ Updatable
                          _ System requires external power source
                          _ Needs shutdown after installation
```

This isn't stock BIOS, but coreboot ported for these laptops:

```
$ fwupdtool hwids
Computer Information
--------------------
BiosVendor: 3mdeb Embedded Systems Consulting
BiosVersion: Dasharo (coreboot+UEFI) v1.0.0
Manufacturer: Notebook
Family: Not Applicable
ProductName: NS50MU
ProductSku: Not Applicable
EnclosureKind: a
BaseboardManufacturer: Notebook
BaseboardProduct: NS50MU

Hardware IDs
------------
not available as 'BiosMajorRelease' unknown
not available as 'BiosMajorRelease' unknown
not available as 'BiosMajorRelease' unknown
{60f53465-e8fc-5122-b79b-f7b03f063037}   <- Manufacturer + Family + ProductName + ProductSku + BaseboardManufacturer + BaseboardProduct
{99a74816-5cd4-5166-b844-df3ef29e7300}   <- Manufacturer + Family + ProductName + ProductSku
{00bb8e01-c72a-5906-aa0e-2279c54ba369}   <- Manufacturer + Family + ProductName
{b05c067a-a96f-591b-9c69-7dda03324e00}   <- Manufacturer + ProductSku + BaseboardManufacturer + BaseboardProduct
{cd9d0da1-e88d-5d36-aae5-2ba60e8d0ed6}   <- Manufacturer + ProductSku
{f42b9241-6cdc-55cb-bb62-393a52752fc8}   <- Manufacturer + ProductName + BaseboardManufacturer + BaseboardProduct
{b7b416f9-55aa-5fd4-bff7-fbaa4143bd92}   <- Manufacturer + ProductName
{b05c067a-a96f-591b-9c69-7dda03324e00}   <- Manufacturer + Family + BaseboardManufacturer + BaseboardProduct
{cd9d0da1-e88d-5d36-aae5-2ba60e8d0ed6}   <- Manufacturer + Family
{a127c2d7-5537-5b72-8bc1-9ccd683f18e6}   <- Manufacturer + EnclosureKind
{774c2cef-39d8-5b40-9477-c220b4fc1f04}   <- Manufacturer + BaseboardManufacturer + BaseboardProduct
{56bb94fc-caf1-5863-9e85-1e449a0c5a01}   <- Manufacturer
```

> and then we need it defined in the plugins/flashrom/README.md file....

You mean describe structure of instance ID there?

SergiiDmytruk commented:
By the way, there is an alternative method for unlocking ME region of these laptops: holding <kbd>Fn + M</kbd> when pressing power on button.

So there is also an option of just giving a user instructions like that before updating ME firmware. This makes changes to SuperIO plugin unnecessary and gets rid of unlocking of EC device to get to ME region, which looks a bit weird.

Which way of unlocking would you prefer?

hughsie commented:
I'm not a fan of the manual action, but I do like the simplicity.

superm1 commented:
From a purely security oriented view, should you really be able to allow to turn off ME from software in an operating system? I feel like at least requiring physical interaction is a feature.

SergiiDmytruk commented:
Changed to manual unlocking to see how it might look. It caused addition of some new code in flashrom plugin to make it possible, so not pure simplification. Changes in superio plugin are rather independent, so both ways of unlocking can even be provided at the same time.

> From a purely security oriented view, should you really be able to allow to turn off ME from software in an operating system? I feel like at least requiring physical interaction is a feature.

An argument, but If kernel isn't locked down, there is a long list of things that can be done to the system besides ME unlocking.

superm1 commented:
>Changed to manual unlocking to see how it might look. It caused addition of some new code in flashrom plugin to make it possible, so not pure simplification

OK, will look through it.

> An argument, but If kernel isn't locked down, there is a long list of things that can be done to the system besides ME unlocking.

Sure; but the ME unlocking is a permanent problem.  So if userspace software can accomplish this, that means malware can accomplish it.

superm1 commented on plugins/flashrom/fu-plugin-flashrom.c:
```diff
@@ -183,11 +271,39 @@ fu_plugin_flashrom_startup(FuPlugin *plugin, GError **error)
+	    fu_flashrom_device_has_fn_m_me_unlock(flashrom_device)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
```
maybe this should be NOTHING_TO_DO?

SergiiDmytruk replied:
Right, thanks.

superm1 commented on plugins/flashrom/fu-plugin-flashrom.c:
```diff
@@ -183,11 +271,39 @@ fu_plugin_flashrom_startup(FuPlugin *plugin, GError **error)
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "\n"
+				    "ME region on Tuxedo laptops should be unlocked manually:\n"
+				    " 1. Power off your laptop\n"
+				    " 2. Press and keep holding Fn + M during the next step\n"
+				    " 3. Press power on button");
```
I suspect @hughsie won't be a fan of hardcoded error strings, this means translators won't be able to help in translations.  But this is a pretty advanced use case.  Maybe it is reasonable.

If the strings stay, presumably you don't want to include the words "Tuxedo" or "laptop" to future proof the string.  That way other models may support it, and other form factors may support it.  You can move tuxdeo and laptop into a comment near by.

SergiiDmytruk replied:
I see only one use of translatable strings in plugins:
https://github.com/fwupd/fwupd/blob/7ec0f11e6925fd59f3a1a8f177b90de54f40a300/plugins/uefi-capsule/fu-plugin-uefi-capsule.c#L447
And it seems to not be used... I'll add `_(` and `TRANSLATORS:` comment as it seems to be supported in plugins.

Another thing is that verbose instructions are given via an error message, which is sort of a misuse, but I didn't see another way of doing it.

hughsie replied:
Na, no translations in plugins please. Ideally it'd be an enumerated fwupdrequest type, but that's hard for this kind of thing.

superm1 reviewed (dismissed):
Sort out the pre-commit problem, and then this approach LGTM overall.
I do think you should remove the unlock interface from the EC too for security of the system, but that is up to you.

SergiiDmytruk commented:
> Sort out the pre-commit problem, and then this approach LGTM overall.

That's just formatting. I'll run `clang-format` when marking PR as ready for review.

> I do think you should remove the unlock interface from the EC too for security of the system, but that is up to you.

I did. Just wanted to mention that it can be brought back if needed.

By the way, I noticed a weird difference between outputs of `fwupdtool get-devices --plugins flashrom` and `fwupdmgr get-devices --show-all-devices`. Output of the first command doesn't have these lines:
```
                    ? Device is locked
                    ? Cryptographic hash verification is available
```
Is this expected or a bug? At first glance, the same code is used in both cases.

superm1 commented:
> That's just formatting. I'll run clang-format when marking PR as ready for review.

OK

>I did. Just wanted to mention that it can be brought back if needed.

Right I see it's gone here; but I think you should remove it from the EC codebase too.  That's of course out of my control though :)

>Is this expected or a bug? At first glance, the same code is used in both cases.

That doesn't sound expected to me.

SergiiDmytruk commented:
> but I think you should remove it from the EC codebase too

That's developed by Clevo, not Tuxedo or 3mdeb :) We're making it possible to update EC using open source tools and also [ported](https://docs.dasharo.com/variants/tuxedo_ibs15/releases/) coreboot to these laptops.

> That doesn't sound expected to me.

It's fine after all. `--plugins` is actually `--plugin-whitelist`, so no plugin interactions on `--plugins flashrom` and result is different.

superm1 commented:
Okay no concerns left for me.

hughsie commented on plugins/flashrom/fu-flashrom-device.h:
```diff
@@ -13,4 +13,10 @@
+fu_flashrom_device_get_region(FuFlashromDevice *self);
+
+gboolean
+fu_flashrom_device_has_fn_m_me_unlock(FuFlashromDevice *self);
```
I don't like this much, can you just do `fu_device_has_private_flag(device, FU_FLASHROM_DEVICE_FLAG_FN_M_ME_UNLOCK);` in the caller please. You can move the `FU_FLASHROM_DEVICE_FLAG_FN_M_ME_UNLOCK` to header if that helps.

hughsie commented on plugins/flashrom/fu-plugin-flashrom.c:
```diff
@@ -143,14 +143,41 @@ fu_plugin_flashrom_device_set_hwids(FuPlugin *plugin, FuDevice *device)
+	/* add this so we can attach all the other quirks */
+	fu_device_add_instance_str(FU_DEVICE(device), "GUID", guid);
+	if (!fu_device_build_instance_id(FU_DEVICE(device), error, "FLASHROM", "GUID", NULL)) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "flashrom failed to add GUID instance id for %s device", region_str);
```
you don't need this, fu_device_build_instance_id will set the error for you.

SergiiDmytruk replied:
Sure, thanks, varargs confused me.

hughsie commented on plugins/flashrom/fu-plugin-flashrom.c:
```diff
@@ -143,14 +143,31 @@ fu_plugin_flashrom_device_set_hwids(FuPlugin *plugin, FuDevice *device)
+	fu_device_add_instance_str(device, "REGION", region_str);
+	fu_device_add_instance_str(device, "VENDOR", vendor);
+	fu_device_add_instance_str(device, "PRODUCT", product);
+	if (!fu_device_build_instance_id(device, NULL, "FLASHROM", "VENDOR", "PRODUCT", "REGION",
+					 NULL)) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "flashrom failed to add instance id for %s device", region_str);
```
just set the error for fu_device_build_instance_id please

SergiiDmytruk commented:
Noticeable change apart from addressing comments: device for ME partition isn't created unless `intel_spi` creates its corresponding device. If depending on `intel_spi` in this way is undesirable, can check for the region via flashrom by creating a context, loading layout from IFD and checking if including `me` region fails (no "is_present" function). Which makes me wonder whether flashrom context should be managed by the plugin instead of device object.

hughsie commented:
> whether flashrom context should be managed by the plugin

Yes, this sounds right.

SergiiDmytruk commented:
Moved context management and also fixed leak of layout. One thing to note: operations depend on the state of the context, so if it's possible to use two devices at the same time, this change shouldn't be accepted.

hughsie commented on plugins/flashrom/fu-flashrom-device.c:
```diff
@@ -23,12 +23,12 @@ struct _FuFlashromDevice {
 	FuIfdRegion region;
 	struct flashrom_flashctx *flashctx;
-	struct flashrom_programmer *flashprog;
+	struct flashrom_layout *layout;
```
Why do we need to store the flashrom_layout in the device? It doesn't seem to me we use it anywhere.

hughsie replied:
Ohh i see, we use it during ->open() to set on the context; ignore me.

hughsie replied:
Nope, we clear it during ->close() -- I think my original point holds.

SergiiDmytruk replied:
I just needed to store it somewhere so it doesn't leak. `flashrom_layout_set()` doesn't transfer ownership:
```c
/**
 * @brief Set the active layout for a flash context.
 *
 * Note: This just sets a pointer. The caller must not release the layout
 *       as long as he uses it through the given flash context.
 *
 * @param flashctx Flash context whose layout will be set.
 * @param layout   Layout to bet set.
 */
void flashrom_layout_set(struct flashrom_flashctx *const flashctx, const struct flashrom_layout *const layout)
{
	flashctx->layout = layout;
}
```

And there is no `flashrom_layout_get()`, so you can't extract it from the context to free later.

hughsie replied:
> flashrom_layout_set() doesn't transfer ownership

Bleh. Okay, thanks for the explanation. Could you put that rationale in the source code please; one line is fine.

hughsie commented on plugins/flashrom/fu-plugin-flashrom.c:
```diff
@@ -142,15 +161,40 @@ fu_plugin_flashrom_device_set_hwids(FuPlugin *plugin, FuDevice *device)
+	fu_device_add_instance_str(device, "REGION", region_str);
+	fu_device_add_instance_str(device, "VENDOR", vendor);
+	fu_device_add_instance_str(device, "PRODUCT", product);
+	if (!fu_device_build_instance_id(device,
+					 error,
+					 "FLASHROM",
+					 "VENDOR",
+					 "PRODUCT",
+					 "REGION",
+					 NULL))
+		return NULL;
```
I think this needs documenting in README.md

hughsie commented on plugins/flashrom/fu-plugin-flashrom.c:
```diff
@@ -205,15 +286,76 @@ fu_plugin_flashrom_startup(FuPlugin *plugin, GError **error)
 }
 
+static gboolean
+fu_plugin_flashrom_unlock(FuPlugin *self, FuDevice *device, GError **error)
```
out of interest; why here and not in fu-flashrom-device.c?

SergiiDmytruk replied:
Unlock operation is defined only at the plugin level, at least I don't see corresponding function for a device. Or do you suggest to create `fu_flashrom_device_unlock()` and call it here?

hughsie replied:
> to create fu_flashrom_device_unlock() and call it here

Yes, I think that's a good idea. Then if we do a FuDevice->unlock() vfunc in the future we can connect it up easily.

hughsie commented on plugins/flashrom/README.md:
```diff
@@ -45,6 +48,16 @@ Internal device uses hardware ID values which are derived from SMBIOS.
+SPI Flash Descriptor), examples:
+
+* `FLASHROM\VENDOR_Notebook&PRODUCT_NS50MU&REGION_bios`
+* `FLASHROM\VENDOR_Notebook&PRODUCT_NS50MU&REGION_me`
```
on the REGION only that is.

SergiiDmytruk replied:
Probably a good idea to not depend on the case here. Updated.
